### PR TITLE
[Scheduler design] The pragmatic approach

### LINKED ISSUE #336: scheduler leaky abstractions in pipelines
Excited by the opportunities that Stable Diffusion has made for use with consumer hardware, I've started working with the code. I was looking to get an idea of what I'd need to do for something like #277 and studying the StableDiffusionPipeline class.

I'm concerned by the use of ` isinstance(self.scheduler, LMSDiscreteScheduler)`. If you can't implement a new Scheduler without meddling with the internal implementation of the Pipeline, that points to some breakdown in the API design. We don't want growing `if isintance` trees that grow with each New Scheduler PR.

It also introduces an ordering dependency between components that weren't otherwise entangled: https://github.com/huggingface/diffusers/blob/e49dd03d2de347caf5f18d04657500412f57103b/src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py#L120-L124

It wasn't obvious to me at first glance, but `scheduler.set_timesteps` modifies `scheduler.sigmas`. So now when working with that code we have to remember that we have to do that scheduler initialization _before_ we finish preparing the initial latents, whereas those were independent operations before.

Another worrying comment is this: https://github.com/huggingface/diffusers/blob/e49dd03d2de347caf5f18d04657500412f57103b/src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py#L126-L130

If "not all schedulers share the same signature," it's much harder to make and use Schedulers.  

## Suggestions

LMSDiscreteScheduler's `step` function seems to want the step number instead of the time. Options:

1. Change the signature of `Scheduler.step` to always include the index as well as the time. (It's readily available and cheap to have on the stack.)
2. Since LMS is linear (right there in the name!), could it take the usual `timestep` value and easily convert it back to its index value?

For DDIMScheduler's extra `eta` parameter:

The `**kwargs` on `StableDiffusionPipeline.__call__` seems to be unused, and `eta` seems to be passed straight through to `scheduler.step` with no other interaction. We could take `eta` out of the formal parameter list and rename `kwargs` to `extra_step_kwargs`.

I haven't yet looked at other uses of Scheduler outside of this one pipeline, but those are my initial thoughts.

### PR DESCRIPTION
**Pull request by anton-l**
This schedulers API redesign addresses concerns raised in #336 and starts to make the schedulers interchangeable without making scheduler class-dependent customizations to pipelines.

1) Now every scheduler contains an `init_noise_sigma` parameter to scale the normal distribution of the initial noise. While it is just `1.0` for DDPM, DDIM and PNDM, it *is* customized for the VE and K-LMS schedulers.
Example usage:
```python
sample = torch.randn(*shape, generator=generator) * self.scheduler.init_noise_sigma
```
2) Every scheduler needs to implement the `scale_model_input(sample, timestep)` (even if it just returns the `sample`) that scales the denoising model's input based on the current `timestep`. The method should be called before every `model()` call.
Example usage:
```python
sample = self.scheduler.scale_model_input(sample, t)
output = model(sample, t)
```
Note: the decision to not make it a base class method is intentional, as suggested by @patrickvonplaten 

Closes #336

### CODE REVIEW

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_img2img.py [hunk 1]
```diff
@@ -222,6 +222,7 @@ def __call__(
         offset = self.scheduler.config.get("steps_offset", 0)
         init_timestep = int(num_inference_steps * strength) + offset
         init_timestep = min(init_timestep, num_inference_steps)
+        # FIXME
         if isinstance(self.scheduler, LMSDiscreteScheduler):
             timesteps = torch.tensor(
                 [num_inference_steps - init_timestep] * batch_size, dtype=torch.long, device=self.device
```
Comment (anton-l) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_img2img.py:228]: The timesteps in LMS are in reverse: ideally need to keep them in a consistent order between schedulers

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_inpaint.py [hunk 2]
```diff
@@ -338,12 +334,12 @@ def __call__(
                 noise_pred = noise_pred_uncond + guidance_scale * (noise_pred_text - noise_pred_uncond)
 
             # compute the previous noisy sample x_t -> x_t-1
+            latents = self.scheduler.step(noise_pred, t, latents, **extra_step_kwargs).prev_sample
             if isinstance(self.scheduler, LMSDiscreteScheduler):
-                latents = self.scheduler.step(noise_pred, t_index, latents, **extra_step_kwargs).prev_sample
                 # masking
+                t_index = t_start + i
                 init_latents_proper = self.scheduler.add_noise(init_latents_orig, noise, torch.LongTensor([t_index]))
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_inpaint.py:341]: Couldn't the `add_noise` function also swallow the complexity here by converting `t` to `t_index` ?

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_inpaint.py [hunk 3]
```diff
@@ -338,12 +334,12 @@ def __call__(
...
             else:
```
Comment (anton-l) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_inpaint.py:342]: Same issue with the timestep order as above

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_inpaint.py (hunk 2 above)
Comment (anton-l) [reply src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_inpaint.py:341]: `add_noise` expects indices instead of timesteps, which are in reverse order (same issue as img2img). Trying to find an elegant solution right now, open to suggestions

**Review by patrickvonplaten — COMMENTED** [review]
Awesome, love this API design - think it's slim and should solve 99% of our problems! Actually maybe no need after all to have a `SchedulerType.CONTINUOUS` then after all in a first step :heart_eyes:

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_inpaint.py (hunk 2 above)
Comment (patrickvonplaten) [reply src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_inpaint.py:341]: Think the `add_noise` method should just be able handle the same input type as `add_noise` of the other schedulers

**Comment by anton-l** [comment]
The slow tests pass :heavy_check_mark:

File: src/diffusers/schedulers/scheduling_lms_discrete.py [hunk 4]
```diff
@@ -206,12 +225,14 @@ def add_noise(
         self,
         original_samples: torch.FloatTensor,
         noise: torch.FloatTensor,
-        timesteps: torch.IntTensor,
+        timesteps: torch.FloatTensor,
     ) -> torch.FloatTensor:
         sigmas = self.sigmas.to(original_samples.device)
+        schedule_timesteps = self.timesteps.to(original_samples.device)
         timesteps = timesteps.to(original_samples.device)
+        step_indices = [(schedule_timesteps == t).nonzero().item() for t in timesteps]
```
[the remark this answers is not in the document]
Comment (patrickvonplaten) [reply src/diffusers/schedulers/scheduling_lms_discrete.py:233]: don't think it's that bad honestly

**Comment by patrickvonplaten** [comment]
Really cool - like this new design a lot! 

Think we can merge this tomorrow :heart_eyes: 

Some final TODOs / suggestions:
- add the design now to all other schedulers and their scripts (or leave as clear TODOs)
- think we could add a test that forces every new scheduler to have:
    - a `init_noise_sigma` variable
    - a `scale_model_input` function
    - a `step` function 
    - => this could ensure that we follow this design in the future. If some schedulers don't follow this design, let's exempt them for this test maybe for now with a big TODO to fix it
- Give `scale_model_input` nice docstrings and make sure it's displayed well in the docs. Also let's maybe add a short comment to DDIM, PNDM and DDPM stating that they don't need model scaling
- Add a big :warning: :warning: to this PR that it's backwards breaking and let's maybe try to make it easy for users to fix their code by:
    - If the first timestep that is passed to LMS is an int and is 0 => then it's very likely to be wrong (let's throw a warning here)
    - If LMS `step` function is used before having called `scale_model_input` it's most likely wrong (let's throw an maybe even an error here)

File: src/diffusers/schedulers/scheduling_lms_discrete.py [hunk 5]
```diff
@@ -177,7 +202,21 @@ def step(
...
+        if not isinstance(timestep, float) and not isinstance(timestep, torch.FloatTensor):
+            warnings.warn(
+                f"`LMSDiscreteScheduler` timesteps must be `float` or `torch.FloatTensor`, not {type(timestep)}. "
+                "Make sure to pass one of the `scheduler.timesteps`"
+            )
+        if not self.is_scale_input_called:
+            warnings.warn(
+                "The `scale_model_input` function should be called before `step` to ensure correct denoising. "
+                "See `StableDiffusionPipeline` for a usage example."
+            )
```
Comment (anton-l) [inline src/diffusers/schedulers/scheduling_lms_discrete.py:214]: This will pop up in existing community pipelines but won't break them like an exception would. The legacy pipelines can continue using the manual scaling code :+1:

File: tests/test_scheduler.py [hunk 6]
```diff
@@ -226,6 +226,27 @@ def recursive_check(tuple_object, dict_object):
 
             recursive_check(outputs_tuple, outputs_dict)
 
+    def test_scheduler_public_api(self):
```
Comment (patrickvonplaten) [inline tests/test_scheduler.py:229]: Very nice!

**Review by patrickvonplaten — APPROVED** [review]
Very nice! Happy to merge and help you update existing notebooks / docs / blog post now:
- Update PR description that states exactly what people have to change if they have written their own custom loop. E.g. If you have been using the K-LMS scheduler, please make sure to do the following:
- If you have been using other schedulers, no need to change anything, but we recomend for generality to always make use of `init_sigma` and `scale_model_input`
- All blog posts
- All notebooks
- All training examples
